File: java/MLAT.java
package kr.ac.postech.monet.psmeasureapp_ble;

import android.util.Log;

import no.uib.cipr.matrix.*;

public class MLAT {
    public class GdescentResult
    {
        // DenseMatrix and DenseVector has double elements by default
        public Vector estimator;
        public DenseMatrix estimator_candidate;
        public DenseVector error;

        public GdescentResult(int n, int dim)
        {
            estimator_candidate = new DenseMatrix(n, dim);
            error = new DenseVector(n);
        }
    }

    private static final int n_trial_default = 100;
    private static final double alpha_default = 0.001;
    private static final double time_threshold_default = 0;

    private int n_trial = n_trial_default;
    private double alpha = alpha_default;
    private double time_threshold = time_threshold_default;

    // Euclidian distance: need to verify
    public static double d(Vector p1, Vector p2)
    {
        //return Distance.Euclidean(p1, p2);

        // to prevent the modification of p1

        DenseVector pp1 = new DenseVector(p1);

        return pp1.add(-1, p2).norm(Vector.Norm.Two); // p1.add(-1, p2) = equivalent to p1-p2
    }

    private void initGdescent()
    {
        this.n_trial = n_trial_default;
        this.alpha = alpha_default;
        this.time_threshold = time_threshold_default;
    }
    public void set_n_trial(int n_)
    {
        this.n_trial = n_;
    }
    public void set_alpha(double alpha_)
    {
        this.alpha = alpha_;
    }
    public void set_time_threshold(double time_threshold_)
    {
        this.time_threshold = time_threshold_;
    }

    public GdescentResult gdescent(Matrix anchors_in, Vector ranges_in, Matrix bounds_in_)
    {
        // As Java doesn't have default parameter value, let n_trial, alpha, and time_threshold as class variable
        // If you want to use customized value of n_trial, alpha or time_threshold, use set_n_trial(), set_alpha() or set_time_threshold() instead

        int n = anchors_in.numRows();
        int dim = anchors_in.numColumns();

        GdescentResult gdescent_result = new GdescentResult(this.n_trial, dim);

        DenseMatrix bounds_in;

        if (bounds_in_ == null)
            bounds_in = new DenseMatrix(1, dim);
        else
            bounds_in = new DenseMatrix(bounds_in_);

        DenseMatrix bounds_temp = StackMatrix(anchors_in, bounds_in);

        /*
        Matrix<double> bounds = DenseMatrix.Build.Dense(2, dim);
        for (int i = 0; i < dim; i++)
        {
            bounds[0, i] = bounds_temp.Column(i).Min();
            bounds[1, i] = bounds_temp.Column(i).Max();
        }
        */
        DenseMatrix bounds = new DenseMatrix(2, dim);
        for (int i = 0; i < bounds_temp.numColumns(); i++)
        {
            double columnMin = 0.0;
            double columnMax = 0.0;

            int columnStartIdx = i * bounds_temp.numRows();
            int columnEndIdx = (i + 1) * bounds_temp.numRows() - 1;

            columnMin = bounds_temp.getData()[columnStartIdx];
            columnMax = bounds_temp.getData()[columnStartIdx];

            for (int j = columnStartIdx + 1; j <= columnEndIdx; j++) {
                if (bounds_temp.getData()[j] < columnMin)
                    columnMin = bounds_temp.getData()[j];
                if (bounds_temp.getData()[j] > columnMax)
                    columnMax = bounds_temp.getData()[j];
            }

            bounds.set(0, i, columnMin);
            bounds.set(1, i, columnMax);
        }

        if (this.time_threshold == 0)
            this.time_threshold = 1.0 / this.n_trial;

        DenseVector ranges = new DenseVector(n);

        for (int i = 0; i < this.n_trial; i++)
        {
            DenseVector estimator0 = new DenseVector(dim);
            for (int j = 0; j < dim; j++)
            {
                estimator0.set(j, Math.random() * (bounds.get(1, j) - bounds.get(0, j)) + bounds.get(0, j));
            }

            DenseVector estimator = new DenseVector(estimator0);

            long startTime = System.currentTimeMillis();
            while (true)
            {
                for (int j = 0; j < n; j++)
                {
                    ranges.set(j, d(getRow(anchors_in, j), estimator));
                }
                double error = d(ranges_in, ranges);

                DenseVector delta = new DenseVector(dim);
                for (int j = 0; j<n; j++)
                {
                    // to prevent estimator unintentionally changed
                    DenseVector estimatorTmp = new DenseVector(estimator);

                    delta.add((ranges_in.get(j) - ranges.get(j)) / ranges.get(j), estimatorTmp.add(-1, getRow(anchors_in, j)));
                }
                delta.scale(2 * this.alpha);

                // to prevent estimator unintentionally changed
                DenseVector estimatorTmp = new DenseVector(estimator);

                Vector estimator_next = estimatorTmp.add(-1, delta);
                for (int j = 0; j < n; j++)
                {
                    ranges.set(j, d(getRow(anchors_in, j), estimator_next));
                }
                double error_next = d(ranges_in, ranges);
                if (error_next < error)
                {
                    estimator = new DenseVector(estimator_next);
                }
                else {
                    gdescent_result.estimator_candidate = setRow(gdescent_result.estimator_candidate, i, estimator);
                    gdescent_result.error.set(i, error);
                    break;
                }
                if (System.currentTimeMillis() - startTime > this.time_threshold)
                {
                    gdescent_result.error.set(i, Double.MAX_VALUE);
                    break;
                }
            }
        }

        // Back to default value
        this.initGdescent();

        return gdescent_result;
    }

    public GdescentResult mlat(Matrix anchors_in, Vector ranges_in, Matrix bounds_in_)
    {
        // As Java doesn't have default parameter value, let n_trial, alpha, and time_threshold as class variable
        // If you want to use customized value of n_trial, alpha or time_threshold, use set_n_trial(), set_alpha() or set_time_threshold() instead

        GdescentResult gdescent_result = this.gdescent(anchors_in, ranges_in, bounds_in_);

        int idx = -1;
        double error = Double.MAX_VALUE;
        for (int i = 0; i < gdescent_result.error.size(); i++)
        {
            if (gdescent_result.error.get(i) < error)
            {
                idx = i;
                error = gdescent_result.error.get(i);
            }
        }

        gdescent_result.estimator = getRow(gdescent_result.estimator_candidate, idx);
        return gdescent_result;
    }

    private static DenseMatrix StackMatrix(Matrix top, Matrix bottom)
    {
        int n1 = top.numRows();
        int n2 = bottom.numRows();

        int dim1 = top.numColumns();
        int dim2 = bottom.numColumns();

        if (dim1 != dim2)
            return null;

        double[] tmp = new double[(n1 + n2) * dim1];

        DenseMatrix top_tmp = new DenseMatrix(top);
        double[] tmp1 = top_tmp.getData();
        for (int i=0; i<tmp1.length; i++) {
            tmp[i + i / n1 * n2] = tmp1[i];
        }

        DenseMatrix bottom_tmp = new DenseMatrix(bottom);
        double[] tmp2 = bottom_tmp.getData();
        for (int i=0; i<tmp2.length; i++) {
            tmp[n1 * (i / n2 + 1) + i] = tmp2[i];
        }

        DenseMatrix result = new DenseMatrix(n1 + n2, dim1, tmp, true);

        return result;
    }

    public static Matrix getMatrix(double[][] matrixDataInRowOrder)
    {
        int numRow = matrixDataInRowOrder.length;
        int numColumn = -1;

        boolean equivalent = true;

        if (numRow > 0)
            numColumn = matrixDataInRowOrder[0].length;

        for (int i = 1; i < numRow; i++) {
            if (matrixDataInRowOrder[i].length != numColumn)
            {
                equivalent = false;
                break;
            }
        }

        if (equivalent == false)
            return null;

        double[] columnOrderData = new double[numRow * numColumn];

        int idx = 0;
        for (int i = 0; i < numColumn; i++) {
            for (int j = 0; j < numRow; j++) {
                columnOrderData[idx] = matrixDataInRowOrder[j][i];
                idx++;
            }
        }

        DenseMatrix result = new DenseMatrix(numRow, numColumn, columnOrderData, true);
        return result;
    }

    public static Vector getRow(Matrix matrix, int row)
    {
        if (row < 0 || row >= matrix.numRows())
            return null;

        DenseVector vector = new DenseVector(matrix.numColumns());

        for (int i = 0; i < matrix.numColumns(); i++)
            vector.set(i, matrix.get(row, i));

        return vector;
    }

    public static DenseMatrix setRow(DenseMatrix matrix, int row, DenseVector vector)
    {
        for (int i = 0; i < vector.size(); i++)
            matrix.set(row, i, vector.get(i));

        return matrix;
    }

    public static String printContents(Matrix matrix)
    {
        String contents = "";

        contents += matrix.getClass().getSimpleName() + " " + matrix.numRows() + "x" + matrix.numColumns() + "-Double";
        contents += "\n";

        for (int i = 0; i < matrix.numRows(); i++)
        {
            for (int j = 0; j < matrix.numColumns(); j++)
                contents += String.format("%.6f", matrix.get(i, j)) + " ";

            contents += "\n";
        }

        return contents;
    }

    public static String printContents(Vector vector)
    {
        String contents = "";

        contents += vector.getClass().getSimpleName() + " " + vector.size() + "-Double";
        contents += "\n";

        for (int i = 0; i < vector.size(); i++)
        {
            contents += String.format("%.6f", vector.get(i)) + "\n";
        }

        return contents;
    }

    /*
    public void demo()
    {
        System.Random random = new SystemRandomSource();

        double W = 9, L = 9, H = 3;
        Matrix<double> anchors = DenseMatrix.OfArray(new double[,] {
            {0, 0, H},
            {W, 0, H},
            {W, L, H},
            {0, L, H}
        });


        // Initial Demo
        DenseMatrix anchors_in = new DenseMatrix(getMatrix(anchors_in_matrix));

        int n = anchors_in.numRows();
        int dim = anchors_in.numColumns();

        DenseMatrix bounds_in = new DenseMatrix(1, dim);

        DenseMatrix bounds_temp = StackMatrix(anchors_in, bounds_in);
        DenseMatrix bounds = new DenseMatrix(2, dim);

        for (int i = 0; i < dim; i++)
        {
            double columnMin = 0.0;
            double columnMax = 0.0;

            int columnStartIdx = i * bounds_temp.numRows();
            int columnEndIdx = (i + 1) * bounds_temp.numRows() - 1;

            columnMin = bounds_temp.getData()[columnStartIdx];
            columnMax = bounds_temp.getData()[columnStartIdx];

            for (int j = columnStartIdx; j <= columnEndIdx; j++) {
                if (bounds_temp.getData()[j] < columnMin)
                    columnMin = bounds_temp.getData()[j];
                if (bounds_temp.getData()[j] > columnMax)
                    columnMax = bounds_temp.getData()[j];
            }

            bounds.set(0, i, columnMin);
            bounds.set(1, i, columnMax);
        }
    }
    */
}

File: js/mlat.js
/*
 * NOTE: requires numeric.js
 */

function new_vector(n, val) {
  var vec = []
  for (var i = 0; i < n; i++) {
    vec.push(val)
  }
  return vec
}

function new_matrix(dimen, val) {
  var mat = []
  for (var i = 0; i < dimen[0]; i++) {
    mat.push(new_vector(dimen[1], val))
  }
  return mat
}

function match_dim(p1, p2) {
  var dim1 = numeric.dim(p1)
  var dim2 = numeric.dim(p2)
  if (!numeric.same(dim1, dim2)) {
    throw "Dimension mismatch"
  }
}

function d(p1, p2) {
  match_dim(p1, p2)
  return numeric.norm2(numeric.sub(p1, p2))
}

class GdescentResult {
  constructor(n, dim) {
    this.estimator = new_vector(dim)
    this.estimator_candidate = new_matrix([n, dim])
    this.error = new_vector(n)
  }
}

function gdescent(anchors_in, ranges_in, bounds_in,
         n_trial = 100, alpha = 0.001, time_threshold = 0) {
  if (!anchors_in
      || numeric.dim(anchors_in).length != 2 /* 2D matrix */
      || numeric.dim(anchors_in)[0] < 3 /* 3 or more anchors */) {
    throw "Coordinates of 3 or more anchors (2D matrix) must be provided"
  }
  var [n, dim] = numeric.dim(anchors_in)
  if (!ranges_in
      || numeric.dim(ranges_in).length != 1 //* 1D vector */
      || ranges_in.length != n /* match with anchors */) {
    throw "Range measurements for the corresponding anchors (1D vector) must be provided"
  }
  var gdescent_result = new GdescentResult(n_trial, dim)

  if (!bounds_in) {
    var bounds_temp = new_matrix([numeric.dim(anchors_in)[0], dim])
  } else {
    if (numeric.dim(bounds_in).length != 2
        || numeric.dim(bounds_in)[1] != dim) {
      throw "Bounds matrix must have the same column size with the Anchors"
    }
    var bounds_temp = new_matrix([numeric.add(numeric.dim(anchors_in),
      numeric.dim(bounds_in))[0],
      dim])
  }
  for (var i = 0; i < n; i++) {
    for (var j = 0; j < dim; j++) {
      bounds_temp[i][j] = anchors_in[i][j]
    }
  }
  if (bounds_in) {
    for (var i = 0; i < numeric.dim(bounds_in)[0]; i++) {
      for (var j = 0; j < dim; j++) {
        bounds_temp[n + i][j] = bounds_in[i][j]
      }
    }
  }
  var bounds = new_matrix([2, dim])
  for (var j = 0; j < dim;j++) {
    bounds[0][j] = bounds_temp[0][j]
    bounds[1][j] = bounds_temp[0][j]
  }
  for (var i = 0; i < numeric.dim(bounds_temp)[0]; i++) {
    for (var j = 0; j < dim; j++) {
      if (bounds_temp[i][j] < bounds[0][j]) {
        bounds[0][j] = bounds_temp[i][j]
      } else if (bounds_temp[i][j] > bounds[1][j]) {
        bounds[1][j] = bounds_temp[i][j]
      }
    }
  }
  console.log('Bounds')
  console.log(bounds)

  if (time_threshold== 0) {
    var time_threshold = 1 / n_trial
  }

  var ranges = new_vector(n)
  for (var i = 0; i < n_trial; i++) {
    var estimator0 = new_vector(dim)
    for (var j = 0; j < dim; j++) {
      estimator0[j] = Math.random() * (bounds[1][j] - bounds[0][j]) + bounds[0][j]
    }
    var estimator = estimator0.slice()

    t0 = performance.now()
    while (true) {
      for (var j = 0; j < n; j++) {
        ranges[j] = d(anchors_in[j], estimator)
      }
      var error = d(ranges_in, ranges)

      var delta = new_vector(dim, 0)
      for (var j = 0; j < n; j++) {
        factor = (ranges_in[j] - ranges[j]) / ranges[j]
        vector = numeric.sub(estimator, anchors_in[j])
        for (var k = 0; k < dim; k++) {
          delta[k] += factor * vector[k]
        }
      }
      for (var j = 0; j < dim; j++) {
        delta[j] *= 2 * alpha
      }

      var estimator_next = numeric.sub(estimator, delta)
      for (var j = 0; j < n; j++) {
        ranges[j] = d(anchors_in[j], estimator_next)
      }
      var error_next = d(ranges_in, ranges)
      if (error_next < error) {
        estimator = estimator_next
      } else {
        gdescent_result.estimator_candidate[i] = estimator.slice()
        gdescent_result.error[i] = error
        break
      }
      if (performance.now() - t0 > time_threshold * 1000) {
        gdescent_result.error[i] = Number.MAX_VALUE
        break
      }
    }
  }
  return gdescent_result
}

function mlat(anchors_in, ranges_in, bounds_in = null,
         n_trial = 100, alpha = 0.001, time_threshold = 0) {
  var gdescent_result = gdescent(anchors_in, ranges_in, bounds_in,
                             n_trial, alpha, time_threshold)

  var idx = -1
  var error = Number.MAX_VALUE
  for (var i = 0; i < gdescent_result.error.length; i++) {
    if (gdescent_result.error[i] < error) {
      idx = i
      error = gdescent_result.error[i]
    }
  }
  gdescent_result.estimator = gdescent_result.estimator_candidate[idx]
  return gdescent_result
}
